### PR TITLE
Issue #408: Add checkout, build and release workflow

### LINKED ISSUE #408: Add auto pypi release and update releasing guidelines
## Subject of the issue

We need to clarify release guidelines. When to release? Conventions? Write in a guide for posterity? Discuss.

### PR DESCRIPTION
**Pull request by zacbaum**
# Description

This GitHub workflow *should* be something we can trigger manually from the Actions tab (when this has been merged) to stage, build and upload the released package.

Before each release, we will need to ensure that setup.py has been version bumped so as to not try and re-upload the same version of the package twice (pypi won't allow that).

Once this is confirmed working, we can update the release guidelines and close #408.

## Type of change

What types of changes does your code introduce to DeepReg? _Put an `x` in the boxes that
apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation Update (fix or improvement on the documentation)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Other (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR.
If you're unsure about any of them, don't hesitate to ask. We're here to help! This is
simply a reminder of what we are going to look for before merging your code._

- [x] I have
      [installed pre-commit](https://deepreg.readthedocs.io/en/latest/contributing/setup.html)
      using `pre-commit install` and formatted all changed files. If you are not
      certain, run `pre-commit run --all-files`.
- [x] My commits' message styles matches
      [our requested structure](https://deepreg.readthedocs.io/en/latest/contributing/commit.html),
      e.g. `Issue #<issue number>: detailed message`.
- [ ] I have updated the
      [change log file](https://github.com/DeepRegNet/DeepReg/blob/main/CHANGELOG.md)
      regarding my changes. - WILL DO WITH NEXT PR
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate) - WILL DO WITH NEXT PR

### CODE REVIEW

**Comment by zacbaum** [comment]
@mathpluscode @YipengHu @NMontanaBrown - please take a look. This (unfortunately) will be something we can only confirm works once merged...! So, we can merge, make sure the button appears in Actions, then when we want to version bump - update `setup.py` and click go.

File: .github/workflows/python-publish.yml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+jobs:
+  deploy:
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v2
+      - name: Set up Python
+        uses: actions/setup-python@v2
+        with:
+          python-version: "3.x"
```
Comment (mathpluscode) [inline .github/workflows/python-publish.yml:18]: should we fix to 3.7?

**Review by mathpluscode — APPROVED** [review]

**Comment by mathpluscode** [comment]
btw.... be aware that DeepReg is now linked with zenodo, do you want me to deactivate the auto-release on zenodo for the moment, so that you can test releasing however you want ;)

**Comment by zacbaum** [comment]
After some further research... Current changes seem to be the best option going forward.

To summarize: `setup.py` has its version set to 0.0.0 - permanently.

To create a build/release cycle, create and publish a new release on GitHub - named with appropriate semver conventions (e.g. v1.2.3). This action should then get triggered, checkout/setup occurs and then the `setup.py` version is updated to match that of the release. Build occurs and is uploaded to PyPI. 

Done! Easy!

@mathpluscode - not sure how this affects the zenodo stuff but feel free to let me know if this causes any issues.

**Comment by mathpluscode** [comment]
> After some further research... Current changes seem to be the best option going forward.
> 
> To summarize: `setup.py` has its version set to 0.0.0 - permanently. No need for manual bumping.
> 
> To create a build/release cycle, create and publish a new release on GitHub - named with appropriate semver conventions (e.g. v1.2.3). This action should then get triggered, checkout/setup occurs and then the `setup.py` version is updated to match that of the release. Build occurs and is uploaded to PyPI.
> 
> Done! Easy!
> 
> @mathpluscode - not sure how this affects the zenodo stuff but feel free to let me know if this causes any issues. (e.g. what triggers auto release on zenodo?)

@zacbaum a release here https://github.com/DeepRegNet/DeepReg/releases

For safety, I deactivated it for now.

Would you mind updating the docs in RTD as well ;) ?

**Comment by zacbaum** [comment]
> > After some further research... Current changes seem to be the best option going forward.
> > To summarize: `setup.py` has its version set to 0.0.0 - permanently. No need for manual bumping.
> > To create a build/release cycle, create and publish a new release on GitHub - named with appropriate semver conventions (e.g. v1.2.3). This action should then get triggered, checkout/setup occurs and then the `setup.py` version is updated to match that of the release. Build occurs and is uploaded to PyPI.
> > Done! Easy!
> > @mathpluscode - not sure how this affects the zenodo stuff but feel free to let me know if this causes any issues. (e.g. what triggers auto release on zenodo?)
> 
> @zacbaum a release here https://github.com/DeepRegNet/DeepReg/releases
> 
> For safety, I deactivated it for now.
> 
> Would you mind updating the docs in RTD as well ;) ?

Ok - that's what should trigger this as well 😄 

Will update the docs when we know it's working...! This last commit should finally fix all things linting (still getting used to using pre-commit between 2 machines...)

**Comment by mathpluscode** [comment]
@zacbaum regarding the pre-commit on windows, I suggest you to open a ticket anyway, some investigation is needed to solve this. I think it's solvable ^^ just need time.

**Comment by zacbaum** [comment]
> @zacbaum regarding the pre-commit on windows, I suggest you open a ticket anyway, some investigation is needed to solve this. I think it's solvable ^^ just need time.

I've started doing it on Linux and just copying over to windows to make commits, but the missing checks now are just user error of me not doing things properly... 😉

**Comment by NMontanaBrown** [comment]
Don't think I'm qualified to review this, tagging @MattClarkson ?

**Comment by YipengHu** [comment]
> Don't think I'm qualified to review this, tagging @MattClarkson ?

Was about to say exactly the same :D

**Review by mathpluscode — APPROVED** [review]

**Comment by mathpluscode** [comment]
@zacbaum let's merge this? @YipengHu can you review it.

Btw I'm wondering why WIP is blocking this issue.

**Review by YipengHu — APPROVED** [review]

**Comment by zacbaum** [comment]
> Btw I'm wondering why WIP is blocking this issue.

@mathpluscode Have updated from main in case that was blocking.

**Pull request merged**